driver/src/orderbook/streamed/state.rs

```diff
@@ -158,11 +177,150 @@ impl State {
+                // received.
+                let fee_token = self.tokens.get_address_by_id(0).unwrap();
+                let balance = self.balances.entry((*submitter, fee_token)).or_default();
+                balance.balance += *burnt_fees;
+                // This looks weird but we need to call self.apply_trade outside of the match to
+                // prevent multiple mut self borrows.
+                let trades = std::mem::replace(trades, Vec::new());
+                self.pending_solution = PendingSolution::AccumulatingTrades(Vec::new());
+                trades
+            }
+            _ => return,
+        };
+        for trade in trades {
+            // Cannot fail because we ensure that the tokens exist when a trade event is
+            // received and that the order exists and doesn't get removed while it is valid.
+            self.apply_trade(
+                trade.owner,
+                trade.order_id,
+                trade.executed_sell_amount,
+                trade.executed_buy_amount,
+                block_batch_id,
+            )
+            .unwrap();
+        }
+    }
+
+    /// Fails if any tokens or order don't exist.
+    fn apply_trade(
+        &mut self,
+        user: UserId,
+        order: OrderId,
+        executed_sell_amount: u128,
+        executed_buy_amount: u128,
+        batch_id: BatchId,
+    ) -> Result<(), Error> {
+        let order = self
+            .orders
+            .get_mut(&(user, order))
+            .ok_or(Error::UnknownOrder)?;
+        let sell_token = self
+            .tokens
+            .get_address_by_id(order.sell_token)
+            .ok_or(Error::UnknownToken)?;
+        let buy_token = self
+            .tokens
+            .get_address_by_id(order.buy_token)
+            .ok_or(Error::UnknownToken)?;
+
+        if order.has_limited_amount() {
+            order.used_amount += executed_sell_amount;
+        }
+        self.apply_balance_from_trade(user, sell_token, executed_sell_amount, batch_id, true);
+        self.apply_balance_from_trade(user, buy_token, executed_buy_amount, batch_id, false);
+        Ok(())
+    }
+
+    fn apply_balance_from_trade(
+        &mut self,
+        user: UserId,
+        token: TokenAddress,
+        executed_amount: u128,
+        batch_id: BatchId,
+        subtract: bool,
+    ) {
+        let balance = self.balances.entry((user, token)).or_default();
+        balance.update_deposit_balance(batch_id);
+        // It possible that we get an underflow here because the user did not have enough
+        // deposited for the sale but in the same batch filled another order that increased their
+        // balance but whose trade event we have not yet processed.
+        balance.balance = if subtract {
+            balance.balance.overflowing_sub(executed_amount.into()).0
+        } else {
+            balance.balance.overflowing_add(executed_amount.into()).0
+        };
```

Comment:
I feel like overflowing should be an error (if it doesn't overflow back the other way), but I don't see an easy way to implement this.